### PR TITLE
  🎨 web-server api: ordering parameters and simplified openapi specs for complex query parameters

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

 [ReDoc](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/ITISFoundation/osparc-simcore/57f2044cb3ffce6f99ab36b415a8139d04f2c3ae/services/web/server/src/simcore_service_webserver/api/v0/openapi.yaml) [Swagger UI](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/ITISFoundation/osparc-simcore/57f2044cb3ffce6f99ab36b415a8139d04f2c3ae/services/web/server/src/simcore_service_webserver/api/v0/openapi.yaml)
Unifies and simplifies ordering queries. For instance, list `folders`  is displayed as
image
from this **coded openapi specification**
```python
@router.get(
    "/folders",
    response_model=Envelope[list[FolderGet]],
)
async def list_folders(
    _q: Annotated[as_query(FolderListQueryParams), Depends()],
):
    ...    
```

#### ♻️ Unified `ordering` query parameters
- SEE `packages/models-library/src/models_library/rest_ordering.py`
- Introduced a factory function to streamline the handling of ordering parameters, ensuring flexibility for customization per case.
- Refactored existing resources (folders, projects, ...) to adopt the unified approach for defining ordering parameters.
- Established a naming convention for models: `$(ResourceName)$(OperationName)OrderQueryParams` (e.g., `FoldersListOrderQueryParams`).

#### ♻️ Simplified handling of complex query parameters (i.e. JSON-encoded)
- Unified the logic with a **single class** for both OpenAPI specification generation and request validation.
    - Includes the helper function `as_query` for ease of use
- Improved implementation in `api/specs/web-server` to reduce complexity and potential errors.. E.g. openapi specs are simply defined using one model per "group" of parameters
- Note that it uses [dependency injection shortcuts](https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/?h=depends%28%29#type-annotation-vs-depends) to expand all fields in the model as (path `_p`/ query `_q`) parameters.

## Related issue/s

- Preparations to include ordering in ITISFoundation/osparc-issues#468

## How to test

Driving test
```
cd pakcages/models-library
make install-dev
pytest -vv tests/test_rest_ordering
```

## Dev-ops

None

### CODE REVIEW

File: api/specs/web-server/_common.py [hunk 1]
```diff
@@ -8,15 +8,70 @@
...
+                # format="json-string" NOTE: we need to get rid of openapi-core in web-server before using this!
+            )
+            if schema_extras:
+                field_schema.update(schema_extras)
+
+    return _Json
+
+
+def as_query(model_class: type[BaseModel]) -> type[BaseModel]:
+    fields = {}
```
Comment (sanderegg) [inline api/specs/web-server/_common.py:39]: this will be a funny migration job to pyV2

File: api/specs/web-server/_projects_crud.py [hunk 2]
```diff
@@ -45,28 +46,32 @@
...
-    response_model=Envelope[TaskGet],
-    summary="Creates a new project or copies an existing one",
-    status_code=status.HTTP_201_CREATED,
-)
-async def create_project(
-    _params: Annotated[ProjectCreateParams, Depends()],
-    _create: ProjectCreateNew | ProjectCopyOverride,
-    x_simcore_user_agent: Annotated[str | None, Header()] = "undefined",
+class _ProjectCreateHeaderParams(BaseModel):
+    x_simcore_user_agent: Annotated[str | None, Header()] = "undefined"
```
Comment (sanderegg) [inline api/specs/web-server/_projects_crud.py:50]: Q: why did you remove the description here?

File: api/specs/web-server/_resource_usage.py [hunk 3]
```diff
@@ -232,46 +163,37 @@ async def update_pricing_plan(
...
     summary="Update detail information about pricing plan",
     tags=["admin"],
 )
 async def update_pricing_unit(
-    pricing_plan_id: PricingPlanId,
-    pricing_unit_id: PricingUnitId,
-    body: UpdatePricingUnitBodyParams,
+    _p: Annotated[PricingUnitGetPathParams, Depends()],
+    _b: UpdatePricingUnitBodyParams,
 ):
```
Comment (sanderegg) [inline api/specs/web-server/_resource_usage.py:193]: minor: actually. I think writing _path, _query and _body is not so long and goes along using long option names ;) but that's your call.

File: packages/models-library/src/models_library/utils/common_validators.py [hunk 4]
```diff
@@ -20,6 +20,10 @@ class MyModel(BaseModel):
 import operator
 from typing import Any
 
+from orjson import JSONDecodeError
```
Comment (sanderegg) [inline packages/models-library/src/models_library/utils/common_validators.py:23]: rust power!

**Review by sanderegg — APPROVED** [review]
Very nice refactoring.
This is the inverse of parse_as_body/header/param/query methods. nice!
Just a few comments and questions.

File: services/web/server/src/simcore_service_webserver/folders/_models.py [hunk 5]
```diff
@@ -36,35 +40,14 @@ class FolderFilters(Filters):
...
+    default=OrderBy(field=IDStr("modified"), direction=OrderDirection.DESC),
+)
 
 
-class FolderListWithJsonStrQueryParams(
-    PageQueryParameters, FolderListSortParams, FiltersQueryParameters[FolderFilters]
+class FoldersListQueryParams(
+    PageQueryParameters, _FolderOrderQueryParams, FiltersQueryParameters[FolderFilters]  # type: ignore[misc, valid-type]
 ):
     folder_id: FolderID | None = Field(
```
Comment (matusdrobuliak66) [inline services/web/server/src/simcore_service_webserver/folders/_models.py:52]: folder_id and workspace_id here can also be None (more specifically, I think the frontend sends it as a "null" string). This option is not reflected in the OpenAPI specs. NOTE: This was likely the case previously as well.

File: api/specs/web-server/_resource_usage.py (hunk 3 above)
Comment (matusdrobuliak66) [reply api/specs/web-server/_resource_usage.py:193]: I also like the long options more

**Review by matusdrobuliak66 — APPROVED** [review]
Thanks so much! It’s really nice. I’ve left a few comments.

File: api/specs/web-server/_common.py (hunk 1 above)
Comment (pcrespov) [reply api/specs/web-server/_common.py:39]: Let's see. i am optimistic here.

**Review by GitHK — APPROVED** [review]

File: api/specs/web-server/_projects_crud.py (hunk 2 above)
Comment (pcrespov) [reply api/specs/web-server/_projects_crud.py:50]: I did not, it was not there. I will take it from

File: services/web/server/src/simcore_service_webserver/folders/_models.py (hunk 5 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/folders/_models.py:52]: But this is already in place, right?

Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/folders/_models.py:56]: clarified offline.
